usr/sbin/p11sak/p11sak.c

```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
-    free(keytype);
-    return ret;
+    return rc;
 }
 
-static CK_RV execute_cmd(CK_SESSION_HANDLE session, CK_SLOT_ID slot, p11sak_cmd cmd,
-        p11sak_kt kt, CK_ULONG keylength, CK_ULONG exponent, char *ECcurve,
-        char *label, char *attr_string, int long_print) {
-    CK_RV ret;
+/**
+ * Delete objects
+ */
+static CK_RV deleteKey(CK_SESSION_HANDLE session, p11sak_kt kt, char *rm_label,
+                       CK_BBOOL *forceAll)
+{
+    CK_ULONG keylength, count;
+    CK_OBJECT_CLASS keyclass;
+    CK_OBJECT_HANDLE hkey;
+    char *keytype = NULL;
+    char *label = NULL;
+    CK_RV rc;
+    char *buff = NULL;
+
+    if (!(buff = malloc(strlen("y")))) {
+        printf("Error allocating space for buff\n");
+        return CKR_HOST_MEMORY;
+    }
+
+    rc = tok_key_list_init(session, kt, label);
+    if (rc != CKR_OK) {
+        printf("Init token key list failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
+        return rc;
+    }
+
+    while (1) {
+        rc = funcs->C_FindObjects(session, &hkey, 1, &count);
+        if (rc != CKR_OK) {
+            printf("C_FindObjects failed (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            return 1;
+        }
+        if (count == 0)
+            break;
+
+        rc = tok_key_get_key_type(session, hkey, &keyclass, &keytype,
+                &keylength);
+        if (rc != CKR_OK) {
+            printf("Invalid key type (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            continue;
+        }
+
+        rc = tok_key_get_label_attr(session, hkey, &label);
+        if (rc != CKR_OK) {
+            printf("Retrieval of label failed (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+        }
+
+        if (strcmp(rm_label, "") != 0) {
+            if (strcmp(rm_label, label) == 0) {
+
+                if (*forceAll) {
+                    buff = strcpy(buff, "y");
+                } else {
+                    rc = getKeyboardInput(buff, label);
+                    if (rc != CKR_OK) {
+                        goto done;
+                    }
+                }
+
+                rc = finalizeDestroyObject(buff, label, &session, &hkey);
+                if (rc != CKR_OK) {
+                    return rc;
+                }
+                else {
+                    break;
+                }
+            }
+        }
+        else {
+            if (*forceAll) {
+                printf("Destroy Object with Label: %s\n", label);
+                rc = funcs->C_DestroyObject(session, hkey);
+                if (rc != CKR_OK) {
+                    printf(
+                            "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                            label, rc, p11_get_ckr(rc));
+                    return rc;
+                }
+                printf("DONE - Destroy Object with Label: %s\n", label);
+            } else {
+                while (1) {
```

Comment:
what is this while loop for? You exit the loop either by return or by break in any case after one iteration.